Mike's changes for kappa

fvitt opened the pull request:
modified:   carma_enums_mod.F90
        modified:   carma_globaer.h
        modified:   carma_types_mod.F90
        modified:   carmaelement_mod.F90
        modified:   carmagroup_mod.F90
        modified:   carmastate_mod.F90
        modified:   rhopart.F90
        modified:   setupbdif.F90
        modified:   setupgrow.F90
        modified:   setupnuc.F90
        modified:   setupvf.F90
        modified:   wetr.F90
        new file:   hygroscopicity.F90

fvitt committed:
Mike's changes for kappa

        modified:   carma_enums_mod.F90
        modified:   carma_globaer.h
        modified:   carma_types_mod.F90
        modified:   carmaelement_mod.F90
        modified:   carmagroup_mod.F90
        modified:   carmastate_mod.F90
        modified:   rhopart.F90
        modified:   setupbdif.F90
        modified:   setupgrow.F90
        modified:   setupnuc.F90
        modified:   setupvf.F90
        modified:   wetr.F90
        new file:   hygroscopicity.F90

drmikemills commented on rhopart.F90:
```diff
@@ -147,25 +150,24 @@ subroutine rhopart(carma, cstate, rc)
 
-        else
-          ! rlow
+        else ! irhswell(igroup) /= I_WTPCT_H2SO4
```
add another if for I_PETTERS, and then another else for fitzgerand and gerber

tilmes replied:
Hi Mike, have you check if this has been added?

drmikemills replied:
Yes, this is added in @fvitt sandbox:

/glade/work/fvitt/camdev/carma_model_dev/src/physics/carma/base/rhopart.F90

drmikemills commented on wetr.F90:
```diff
@@ -38,6 +41,7 @@ subroutine getwetr(carma, igroup, rh, rdry, rwet, rhopdry, rhopwet, rc, h2o_mass
     real(kind=f), intent(in), optional   :: h2o_mass!! water vapor mass concentration (g/cm3)
     real(kind=f), intent(in), optional   :: h2o_vp  !! water eq. vaper pressure (dynes/cm2)    
     real(kind=f), intent(in), optional   :: temp    !! temperature [K]
+    real(kind=f), intent(in), optional   :: kappa   !! weight percent [%]
```
kappa is hygroscopicity parameter, not weight percent

tilmes replied:
Hi Mike, have you checked if this has been changed?

drmikemills replied:
This has been updated.

drmikemills commented on wetr.F90:
```diff
@@ -197,44 +202,60 @@ subroutine getwetr(carma, igroup, rh, rdry, rwet, rhopdry, rhopwet, rc, h2o_mass
+
+      ! Mixed aerosol paremeterization (Pengfei Yu et al., JAMES, 2015) based on
+      ! Petters and Kreidenweis (ACP, 2007) hygroscopicity parameter kappa
+      if (irhswell(igroup) == I_KAPPA) then 
```
I_PETTERS

drmikemills replied:
This is done.

drmikemills commented on wetr.F90:
```diff
@@ -197,44 +202,60 @@ subroutine getwetr(carma, igroup, rh, rdry, rwet, rhopdry, rhopwet, rc, h2o_mass
+	    rhopwet = rhopdry
+	else ! temp > 190
+	    rwet = rdry * (1._f + humidity*kappa/(1._f-humidity))**(1._f/3._f)
+	    r_ratio  = (rdry / rwet)**3
```
**3.f

drmikemills replied:
It looks like this has not been changed from **3 to **3.f

tilmes replied:
is done

fvitt commented on hygroscopicity.F90:
```diff
@@ -0,0 +1,71 @@
+  
+  1 format('hygroscopicity::ibin=',i4,',core mass=',e10.3,',shell mass=',e10.3,',hygroscopicity=',f9.4)
+              
+  hygro(:NZ,:NBIN,:NGROUP) = 0.5_f ! default
```
initialize to `-huge(1.f)`

drmikemills replied:
Done.

fvitt reviewed (commented):
@drmikemills, make changes to kappa_dev branch.  Please do not change main branch.  I will merge the changes into main when we are ready.
Thanks.

fvitt committed:
Mike's changes after first code review

        modified:   carma_enums_mod.F90
        modified:   carma_globaer.h
        modified:   carma_types_mod.F90
        modified:   carmagroup_mod.F90
        modified:   carmastate_mod.F90
        modified:   hygroscopicity.F90
        modified:   rhopart.F90
        modified:   wetr.F90

fvitt committed:
add option to get dry particle density in CARMASTATE_GetBin

        modified:   carmastate_mod.F90

fvitt committed:
misc fixes

fvitt committed:
coremass checks

        new file:   coremasscheck.F90
        modified:   carmastate_mod.F90
        modified:   coremasscheck.F90
        modified:   hygroscopicity.F90
        modified:   microfast.F90
        modified:   newstate.F90
        modified:   newstate_calc.F90

fvitt committed:
Simone's change

        modified:   prestep.F90

fvitt committed:
fix continuation lines

        modified:   hygroscopicity.F90

zywshoon commented on microfast.F90:
```diff
@@ -131,6 +136,9 @@ subroutine microfast(carma, cstate, iz, scale_threshold, rc)
     enddo
   enddo
 
+  call coremasscheck( carma, cstate, iz, .true.,.false.,.false., "AfterPsolve", rc )
```
I don't think we need to check mass after psolve

cbardeen replied:
I guess you only know by testing, but I think there should be options controlled by the CARMA object to have these report. This would allow you to see where fixes are happening when you run the code, but shouldn't happen normally.

zywshoon replied:
@fvitt @cbardeen 
I assume Chuck suggest in CARMA:
if(DEBUG)
call coremasscheck( carma, cstate, iz, .true.,.true.,.true., "AfterPsolve", rc )
else
call coremasscheck( carma, cstate, iz, .true.,.false.,.false., "AfterPsolve", rc )
end if

And I suggest  delete this call after psolve.

cbardeen replied:
The ones that are fixing and not reporting, I think you always have to have, but whether they report or not should be controllable by a new flag. For exmaple

call coremasscheck( carma, cstate, iz, .true.,mass_checks_report_all,.false., "AfterPsolve", rc )

In the part of the code where you aren't fixing and just checking, I think you want to either only do those for CARMA_DEBUG or add another flag.

if (mass_checks_always) then call coremasscheck( carma, cstate, iz, .false.,true,.true., "elswhere", rc )

tilmes replied:
@zywshoon

zywshoon replied:
done

zywshoon commented on prestep.F90:
```diff
@@ -58,15 +58,13 @@ subroutine prestep(carma, cstate, rc)
           ! substep the gas unless both of these are true. This might run into
           ! trouble if d_t is large and negative.
-!          if (d_gc(iz, igas) < 0._f) then
-          if ((d_gc(iz, igas) < 0._f) .and. ((supsatiold(iz, igas) > 0._f) &
```
we may want to keep this two lines as comments here since people may want to use this one instead sometimes.

cbardeen replied:
Maybe, but it is dangerous to leave code around that isn't used and isn't used for a good reason (i.e. it crashes the model).

zywshoon replied:
make sense, agree with Chuck.

tilmes replied:
is already done

zywshoon commented on prestep.F90:
```diff
@@ -58,15 +58,13 @@ subroutine prestep(carma, cstate, rc)
           ! trouble if d_t is large and negative.
-!          if (d_gc(iz, igas) < 0._f) then
-          if ((d_gc(iz, igas) < 0._f) .and. ((supsatiold(iz, igas) > 0._f) &
-               .or. (supsati(iz, igas) > 0._f))) then
```
and this line

cbardeen replied:
Wasn't the change in the other direction (i.e. to simplify these and ignore the supersat limits)? I am a little confused with what we started with and what we changed here.

tilmes replied:
is done

zywshoon reviewed (changes requested):
have two comments for minor changes.

fvitt commented:
We will need a summary of the changes for the ChangeLog file.

zywshoon commented:
> We will need a summary of the changes for the ChangeLog file.

How about:
Add hygroscopicity subroutines. Add core mass check to ensure the first element of the group doesn't go negative. Add the option that particle neutralization is controlled by the ambiance gas.

cbardeen commented:
Francis,

I am trying to comment on this, but how do I do it. The git documentation I see online says a blue + is supposed to appear when I hover over the line to comment on a line, but I don’t see that.

Thanks,

Chuck

> On Jan 7, 2021, at 10:48 AM, fvitt ***@***.***> wrote:
> 
> 
> @fvitt <https://github.com/fvitt> requested your review on: #1 <https://github.com/ESCOMP/CARMA_base/pull/1> Mike's changes for kappa.
> 
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub <https://github.com/ESCOMP/CARMA_base/pull/1#event-4180996497>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AFLRNREW2FT7HU6AD4VRCBTSYXXX3ANCNFSM4VAFF5UQ>.
>

fvitt commented:
> Francis, I am trying to comment on this, but how do I do it. The git documentation I see online says a blue + is supposed to appear when I hover over the line to comment on a line, but I don’t see that. Thanks, Chuck

In the "Files changed" page, place your curser the right of the line numbers.  It might be easier to demonstrate over video chat.  Let me know.

cbardeen commented:
That’s what I tried to do and it didn’t show anything. Tried both Safari and Chrome. Set up a meeting.



> On Aug 19, 2021, at 9:32 AM, fvitt ***@***.***> wrote:
> 
> 
> Francis, I am trying to comment on this, but how do I do it. The git documentation I see online says a blue + is supposed to appear when I hover over the line to comment on a line, but I don’t see that. Thanks, Chuck
> 
> In the "Files changed" page, place your curser the right of the line numbers. It might be easier to demonstrate over video chat. Let me know.
> 
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub <https://github.com/ESCOMP/CARMA_base/pull/1#issuecomment-902013707>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AFLRNRDQUCZITNT3FKBRRJ3T5UPYJANCNFSM4VAFF5UQ>.
> Triage notifications on the go with GitHub Mobile for iOS <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android <https://play.google.com/store/apps/details?id=com.github.android&utm_campaign=notification-email>.
>

cbardeen commented on carma_types_mod.F90:
```diff
@@ -683,6 +684,7 @@ module carma_types_mod
     !   too_big     .true. if cores are larger than largest CN
     !   nuc_small   .true. if cores are smaller than smallest nucleated CN
     !   rlprod      Latent heat production (per substep) (K/s)
+    !   kappa       Hygroscopicity parameter
```
Should say kappahygro not kappa

drmikemills replied:
okay

tilmes replied:
is done

cbardeen commented on carmaelement_mod.F90:
```diff
@@ -38,7 +38,7 @@ module CARMAELEMENT_mod
   !! @see CARMAELEMENT_Destroy
  subroutine CARMAELEMENT_Create(carma, ielement, igroup, name, rho, itype, icomposition, rc, &
-              shortname, isolute, rhobin, arat)
+              shortname, isolute, rhobin, arat, kappaElement)
```
This is part of the Element object, so the kappa is by defintion the kappa of the element. Similar to other fields, should just call the parameter kappa.

drmikemills replied:
okay

tilmes replied:
@fvitt  will test this

cbardeen commented on carmaelement_mod.F90:
```diff
@@ -261,7 +266,7 @@ subroutine CARMAELEMENT_Print(carma, ielement, rc)
       write(carma%f_LUNOPRT,*) "    icomposition  : ", icomposition
       write(carma%f_LUNOPRT,*) "    isolute       : ", isolute
```
Could/Shoud add kappa to the print.

drmikemills replied:
okay

tilmes replied:
is done

tilmes replied:
@fvitt needs to actually make a local variable...

cbardeen commented on coremasscheck.F90:
```diff
@@ -0,0 +1,66 @@
+
+  real(kind=f), parameter :: thresh  = 1e-10_f     ! relative threshold for core mass and kappa checking
+  real(kind=f), parameter :: thresh1 = 1._f+thresh
+
```
If you are going to have thresholds, the constants should be in carma_constants since they are used in other places (e.g. hygroscopicity). Not sure wy you need a threshold nor want one that is greater than one.

tilmes replied:
is done (code change from Yunquian)

cbardeen commented on coremasscheck.F90:
```diff
@@ -0,0 +1,66 @@
+           end do ! i = 1, ncore(igroup)
+           if (coremass > pc(iz, ibin, iepart) * rmass(ibin, igroup) * thresh1) then
+              if(fixcoremass)then
+                 pc(iz, ibin, iepart) = coremass/rmass(ibin,igroup) + SMALL_PC
```
Why use two different constants? If using a threhold I think you should make sure that the results is bigger than a threshold not smallpc.

drmikemills replied:
@zywshoon

zywshoon replied:
@cbardeen this one is to add some small values to total mass so it is larger than coremass, so I just add a very small value, not the round-up value.

tilmes replied:
is done (code change by Yunqian)

cbardeen replied:
What the code is showing me is that it is still SMALL_PC and not the threshold fraction. Doesn't seem done to me.

tilmes replied:
Chuck, it looks like to me in the code that , if coremass > the total mass
and 
1. if core mass smaller the total mass plus a round off error then we increase the total mass while adding a small value, this is done in any case.
2. if coremass however is larger the total mass and larger than the total mass plus the round-off error, then either we fix this with adding a small value to the total mass, or we print and error (depending on the setting).

I think, this is what we had discussed, the small value is added as Yunqian has described on August 23rd. It that not correct?

tilmes replied:
We agreed to change the SMALL_PC value to "almost_zero" value. SMALL_PC could be removed altogether, however, Yunqian is in favor or adding a very tiny value to have sulfate, not zero, which may be important for adding the nitrate model later.

fvitt replied:
Replace SMALL_PC with ALMOST_ZERO

fvitt replied:
ALMOST_ZERO is now used

cbardeen commented on coremasscheck.F90:
```diff
@@ -0,0 +1,66 @@
+              icore = icorelem(i,igroup)
+              coremass = coremass + pc(iz, ibin, icore)
+           end do ! i = 1, ncore(igroup)
+           if (coremass > pc(iz, ibin, iepart) * rmass(ibin, igroup) * thresh1) then
```
This seems like it will allow cases where coremass is greater than pc by a little, which could generate negative values for the concentration element. Seems like you should apply a threshold the other way if at all to make sure you never get negative values for the concentration element.

drmikemills replied:
@zywshoon

zywshoon replied:
@cbardeen  This is because if core mass is greater than the total by a little bit (i.e. caused by the round-up ), we don't need to fix it. and the round up I set it to 1e-10.

cbardeen replied:
But I think you do need to fix it, since other people that don't check and just do pc * rmass - sum(coremass) will find a negative number that will probably not do the right thing in their calculation if it bigger by up to 1e-10. I also think that 1e-10 is larger than a round off error.

tilmes replied:
@cbardeen Are you suggesting that we reduce the number 1e-10 to a smaller value?
Or are you suggesting to fix always when coremass > pc(iz, ibin, iepart) * rmass(ibin, igroup) ?

tilmes replied:
is done (code change by Yunqian)

cbardeen replied:
Don't think this is right. Should always fix if core mass > PC, but can fix PC to a value that is threshold larger than core mass to keep it from failing again later.

cbardeen commented on hygroscopicity.F90:
```diff
@@ -0,0 +1,90 @@
+              kappahygro(z,ibin,igroup) = kappahygro(z,ibin,igroup) + pc(z,ibin,icore) * kappaelem(icore)
+            end do ! i = 1, ncore(igroup)
+
+            if (coremass.gt.pc(z, ibin, iepart) * rmass(ibin, igroup) * thresh1) then
+              write(LUNOPRT,*) "hygro59: z,ibin,pc,rmass,coremass", &
+                z,ibin,pc(z, ibin, iepart),rmass(ibin, igroup),coremass
+              rc=RC_ERROR
+              return
+            end if
+
+            if (coremass.gt.pc(z, ibin, iepart) * rmass(ibin, igroup)) then
+              !Divide by total mass of all cores in the bin to normalize:
+              kappahygro(z,ibin,igroup) = kappahygro(z,ibin,igroup) / coremass
+            else
```
Why the two cases for coremass? Should we just call the coremass check with the fix before entering this loop and then assume everything is set up correctly in this routine? Would make this code simpler and move the checking/repair to one place that will hopefully get removed in the future.

drmikemills replied:
I was not aware of the coremasscheck subroutine that @zywshoon has written.

tilmes replied:
I agree, we should just run the coremass check!

tilmes replied:
@drmikemills  use the current sandbox and remove coremass check here and test

zywshoon replied:
deleted, done

cbardeen replied:
Still seeing the two cases. Perhaps I am not seeing the right versions of these files, but am looking at what GitHub is showing me.

fvitt replied:
@cbardeen :
What you see here is outdated.  Look at the code under the "Files changed" tab.

cbardeen commented on microfast.F90:
```diff
@@ -47,11 +47,13 @@ subroutine microfast(carma, cstate, iz, scale_threshold, rc)
   if (rc < RC_OK) return
-  
+
+  call coremasscheck( carma, cstate, iz, .false.,.true.,.true., "AfterZeromicro", rc )
```
Why are there calls like this all the time in the code? If you are not going to fix, then this seems like it should only be on for debug. Should probably control this from the CARMA object.

cbardeen replied:
P.S. This occurs in many places, but I won't highlight them all.

drmikemills replied:
@zywshoon

tilmes replied:
Suggest to put a control around this flag. Add a flag in the CARMA object -> could use the DEBUG flag.
@zywshoon

zywshoon replied:
debug flag added , done

cbardeen commented on pheat.F90:
```diff
@@ -211,6 +211,13 @@ subroutine pheat(carma, cstate, iz, igroup, iepart, ibin, igas, dmdt, rc)
                     * rhoelem(ibin,iepart) / 2._f / dtime, &
                    dmdt)
       end if
+    elseif(neutral_volfrc(igroup) .eq. -1._f)then
+      ! When the particle neutralization is not controled by the core mass, 
+      ! instead it is controled by ambience gas
+      ! for example, in trop_strat model, sulfate in mixed particle is neutralized
+      ! by NH4. In this case, we set the neutral_volfrc = -1 for the group
+      dmdt = max((pvap * (ss + 1._f)) * g0, dmdt)
+      !write(*,*) "igroup",igroup," in pheat.F90:neutral_volfrc"
```
Need to think about this a little. I think What this is doing is saying that gas can't leave the particle, but it can go on the particle. The comment here isn't very clear to me about what this is trying to do. How is the ambient gas controlling the evaporation? This just seems like it is shutting it off.

fvitt replied:
Might put this change in a separate pull request?

tilmes replied:
Yes,  lets leave this out here and do a new pull request afterwards

tilmes replied:
@zywshoon  change comment: The sulfate does not evaporate in the mixed particles, because it is assumed that it is neutralized by ammonia.
(@drmikemills need to check if this is a problem in the WACCM model, test this without the change, Yunqian is providing Mike the change to test just applying this in the lower stratosphere.
(@fvitt  keep in the sandbox)

zywshoon replied:
done

cbardeen commented on wetr.F90:
```diff
@@ -197,44 +202,60 @@ subroutine getwetr(carma, igroup, rh, rdry, rwet, rhopdry, rhopwet, rc, h2o_mass
+
+      ! Mixed aerosol paremeterization (Pengfei Yu et al., JAMES, 2015) based on
+      ! Petters and Kreidenweis (ACP, 2007) hygroscopicity parameter kappa
+      if (irhswell(igroup) == I_PETTERS) then 
+	if (temp .le. 190._f) then
+	    rwet = rdry
+	    rhopwet = rhopdry
+	else ! temp > 190
+	    rwet = rdry * (1._f + humidity*kappa/(1._f-humidity))**(1._f/3._f)
+	    r_ratio  = (rdry / rwet)**3
```
Have you plotted what this looks like at the 190 K boundary? Does it transition smoothly? Would it be better to use the value for 190 for everything less that 190 rather than returning the dry values?

drmikemills replied:
This is a valid concern. I have not plotted this, just copied it from Pengfei. It seems like it should be discontinuous, which is a problem. Your suggested fix is not straightforward, however, because temperature affects this indirectly via humidity and kappa, which are calculated elsewhere in the code.

cbardeen replied:
Looks like temperature and water are provided as optional parameters, so it seems like you could recalculate humidity (or crudely scale it by the some exponetial of the temperature ratio or something like that. Not sure about kappa. It is mostly the sum of the composition. What is the temperature dependence on kappa?

tilmes replied:
@drmikemills kappa check

fvitt replied:
Code changed so that when colder than 190K use relative humidity of 190K

fvitt committed:
Changes for the github reviews

        modified:   ChangeLog
        modified:   carma_types_mod.F90
        modified:   carmaelement_mod.F90
        modified:   carmastate_mod.F90
        modified:   coremasscheck.F90
        modified:   wetr.F90

fvitt committed:
More changes for the github reviews

        modified:   carma_globaer.h
        modified:   carma_mod.F90
        modified:   carma_types_mod.F90
        modified:   hygroscopicity.F90
        modified:   microfast.F90
        modified:   pheat.F90

fvitt committed:
Yunqian's changes for wet radius

        modified:   rhopart.F90
        modified:   wetr.F90

fvitt committed:
initialize errabs to zero in fractal_meanfield

        modified:   fractal_meanfield_mod.F90

fvitt committed:
update to latest in CARMA_base main

fvitt committed:
correct ChangeLog

fvitt committed:
supply proper optional args for PETTERS hygroscopicity

        modified:   rhopart.F90
        modified:   wetr.F90

tilmes commented on wetr.F90:
```diff
@@ -36,16 +39,24 @@ subroutine getwetr(carma, igroup, rh, rdry, rwet, rhopdry, rhopwet, rc, h2o_mass
     real(kind=f)            :: rkelvinH2O_a, rkelvinH2O_b, rkelvinH2O, h2o_kelv
-        
+    real(kind=f)            :: rvap,rh190
+
```
@zywshoon  Here we want a comment to describe the particular parameterization of vapor pressure.

zywshoon replied:
@fvitt 
! pvap190h2o uses the vaporp_h2o_murphy2005 equation to calculate the vapor pressure at 190 K for liquid water

fvitt replied:
The comment was added in the code

fvitt committed:
replace SMALL_PC with ALMOST_ZERO; add comment re vap press at 190K

        modified:   coremasscheck.F90
        modified:   wetr.F90

fvitt committed:
update ChangeLog
